### ISSUE TITLE
Custom BatchNormalization layer in ResNet

### ISSUE THREAD

**Issue by rohith513** [keras-team/keras#13105]
**System information**  
- Have I written custom code (as opposed to using example directory):  
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  Arch Linux
- TensorFlow backend (yes / no):  yes
- TensorFlow version:  1.13.1
- Keras version:  2.1.6
- Python version:  3.7
- CUDA/cuDNN version:  10
- GPU model and memory:  GTX 1070

Hello, I am using Faster-RCNN with ResNet50 as base network for object detection. It uses a custom BatchNormalization layer named 'FixedBatchNormalization' in ResNet. I want to know what's the difference between this custom layer and the official BatchNormalization layer in keras. I know that there is an ongoing issue with fine-tuning ResNet when freezing layers. So does the 'FixedBatchNormalization' in this repo solve that issue and does this anyway affect the training and evaluating?

And this layer works only with keras version 2.1.6.
Link to the Faster-RCNN repo:
[https://github.com/kbardool/keras-frcnn](url)

**FixedBatchNormalization code:**

`from keras.engine import Layer, InputSpec
from keras import initializers, regularizers
from keras import backend as K

class FixedBatchNormalization(Layer):

    def __init__(self, epsilon=1e-3, axis=-1,
                 weights=None, beta_init='zero', gamma_init='one',
                 gamma_regularizer=None, beta_regularizer=None, **kwargs):

        self.supports_masking = True
        self.beta_init = initializers.get(beta_init)
        self.gamma_init = initializers.get(gamma_init)
        self.epsilon = epsilon
        self.axis = axis
        self.gamma_regularizer = regularizers.get(gamma_regularizer)
        self.beta_regularizer = regularizers.get(beta_regularizer)
        self.initial_weights = weights
        super(FixedBatchNormalization, self).__init__(**kwargs)

    def build(self, input_shape):
        self.input_spec = [InputSpec(shape=input_shape)]
        shape = (input_shape[self.axis],)

        self.gamma = self.add_weight(shape,
                                     initializer=self.gamma_init,
                                     regularizer=self.gamma_regularizer,
                                     name='{}_gamma'.format(self.name),
                                     trainable=False)
        self.beta = self.add_weight(shape,
                                    initializer=self.beta_init,
                                    regularizer=self.beta_regularizer,
                                    name='{}_beta'.format(self.name),
                                    trainable=False)
        self.running_mean = self.add_weight(shape, initializer='zero',
                                            name='{}_running_mean'.format(self.name),
                                            trainable=False)
        self.running_std = self.add_weight(shape, initializer='one',
                                           name='{}_running_std'.format(self.name),
                                           trainable=False)

        if self.initial_weights is not None:
            self.set_weights(self.initial_weights)
            del self.initial_weights

        self.built = True

    def call(self, x, mask=None):

        assert self.built, 'Layer must be built before being called'
        input_shape = K.int_shape(x)

        reduction_axes = list(range(len(input_shape)))
        del reduction_axes[self.axis]
        broadcast_shape = [1] * len(input_shape)
        broadcast_shape[self.axis] = input_shape[self.axis]

        if sorted(reduction_axes) == range(K.ndim(x))[:-1]:
            x_normed = K.batch_normalization(
                x, self.running_mean, self.running_std,
                self.beta, self.gamma,
                epsilon=self.epsilon)
        else:
            # need broadcasting
            broadcast_running_mean = K.reshape(self.running_mean, broadcast_shape)
            broadcast_running_std = K.reshape(self.running_std, broadcast_shape)
            broadcast_beta = K.reshape(self.beta, broadcast_shape)
            broadcast_gamma = K.reshape(self.gamma, broadcast_shape)
            x_normed = K.batch_normalization(
                x, broadcast_running_mean, broadcast_running_std,
                broadcast_beta, broadcast_gamma,
                epsilon=self.epsilon)

        return x_normed

    def get_config(self):
        config = {'epsilon': self.epsilon,
                  'axis': self.axis,
                  'gamma_regularizer': self.gamma_regularizer.get_config() if self.gamma_regularizer else None,
                  'beta_regularizer': self.beta_regularizer.get_config() if self.beta_regularizer else None}
        base_config = super(FixedBatchNormalization, self).get_config()
        return dict(list(base_config.items()) + list(config.items()))`

**Comment by rohith513** [comment]
Hi @datumbox, I saw your posts on Batchnormalization. But I couldn't get to any conclusion. How is this implementation different from yours?

**Comment by rohith513** [comment]
@datumbox @fchollet Can someone please tell me how to fine-tune ResNet50 here? I want to freeze all Conv blocks except the last one and training the remaining. What is the right way to freeze batchnorm layers so that my validation loss and accuracy improves?

`x = BatchNormalization(axis=3, name=bn_name_base + '2a', trainable = False)(x)`
(or)
`x = BatchNormalization(axis=3, name=bn_name_base + '2a', trainable =False)(x, training = False)`

Which of the above implementation is correct?
